packages/@aws-cdk/aws-iot-actions/lib/sns-topic-action.ts

```diff
@@ -0,0 +1,78 @@
+import * as iam from '@aws-cdk/aws-iam';
+import * as iot from '@aws-cdk/aws-iot';
+import * as sns from '@aws-cdk/aws-sns';
+import { CommonActionProps } from '.';
+import { singletonActionRole } from './private/role';
+
+/**
+ * SNS topic action message format options.
+ */
+export enum SnsActionMessageFormat {
+  /**
+   * RAW message format.
+   */
+  RAW = 'RAW',
+  /**
+   * JSON message format.
+   */
+  JSON = 'JSON'
+}
+
+/**
+ * Configuration options for the SNS topic action.
+ */
+export interface SnsTopicActionProps extends CommonActionProps {
+
+  /**
+   * The message format of the message to publish.
+   *
+   * SNS uses this setting to determine if the payload should be parsed and relevant platform-specific bits of the payload should be extracted.
+   * @see https://docs.aws.amazon.com/sns/latest/dg/sns-message-and-json-formats.html
+   *
+   * @default SnsActionMessageFormat.RAW
+   */
+  readonly messageFormat?: SnsActionMessageFormat;
+}
+
+/**
+ * The action to write the data from an MQTT message to an Amazon SNS topic.
+ *
+ * @see https://docs.aws.amazon.com/iot/latest/developerguide/sns-rule-action.html
+ */
+export class SnsTopicAction implements iot.IAction {
+  private readonly role?: iam.IRole;
+  private readonly topic: sns.ITopic;
+  private readonly messageFormat?: SnsActionMessageFormat;
+
+  /**
+   * @param topic The Amazon SNS topic to publish data on. Must not be a FIFO topic.
+   * @param props Properties to configure the action.
+   */
+  constructor(topic: sns.ITopic, props: SnsTopicActionProps = {}) {
+    if (topic.topicName.endsWith('.fifo')) {
```

Comment:
Discussion probably needed here. 

On the implementation: I don't like having to check for the ".fifo" suffix directly, but `ITopic` doesn't expose anything to check if it is FIFO (unlike `IQueue.fifo`). However, since the suffix is mandated by the service this may be adequate until a better option comes available.

On whether or not the error should even be thrown: The [docs](https://docs.aws.amazon.com/iot/latest/developerguide/sns-rule-action.html) for this action clearly state that FIFO topics are not supported due to the distributed nature of the rules engine. That said, there isn't any validation done by the Rules service to reject a configuration with a FIFO topic. I ran a test with a FIFO queue as the target and received an error in the IoT logs `Invalid parameter: The MessageGroupId parameter is required for FIFO topics (Service: AmazonSNS; Status Code: 400; Error Code: InvalidParameter;`. So I'll argue that throwing this error is likely to save at least a couple people from headache and wasted time debugging a bad configuration.

Reply:
It's also worth noting that the current SQS action implementation does not perform this check, even though [the same FIFO limitation applies](https://docs.aws.amazon.com/iot/latest/developerguide/sqs-rule-action.html). If decide this check is justified we might want to create a ticket to apply the same in the SQS action.

Reply:
I'll second @yamatatsu's comment, this is some excellent detective work @AdamVD! I agree with everything you said.

One subtlety here. Note that `topicName` could be something that CDK calls a [Token](https://docs.aws.amazon.com/cdk/v2/guide/tokens.html). For example, in your integ tests, notice that `tokenName` resolves to `{ Ref: 'MyTopic86869434' }`.

Because of that, it would be a good idea to call the [`Token.isUnresolved()` method](https://docs.aws.amazon.com/cdk/api/v1/docs/@aws-cdk_core.Token.html#static-iswbrunresolvedobj), and only do the check if that method returns `false` (otherwise, we're comparing against the encoded Token, which doesn't make too much sense).

Also, if we need `ITopic` to have a `fifo` property, like `IQueue` does - let's do it! If this is a small change in the SNS module, I'm fine adding it to this PR, otherwise, let's create an issue in the backlog for it, and follow-up after this PR gets merged.

Let me know what you think @AdamVD!

Reply:
Thanks @skinny85 this is a great call out, the Token concept is new new me and I don't yet understand exactly when it is present. However, I wonder if #18740 makes it so that the name of a FIFO topic is always resolved. Regardless, this implementation shouldn't be dependent on a temporary shortcoming of another module, so I'll look into expanding `ITopic` with a `fifo` property as a more robust solution.